sympy/solvers/bivariate.py

```diff
@@ -199,17 +222,60 @@ def _solve_lambert(f, symbol, gens):
     nrhs, lhs = f.as_independent(symbol, as_Add=True)
     rhs = -nrhs
 
+    even_degrees = [i for i in power_list(lhs) if i%2 == 0 and i != 0]
+    t = Dummy('t', **symbol.assumptions0)
+
     lamcheck = [tmp for tmp in gens
                 if (tmp.func in [exp, log] or
                 (tmp.is_Pow and symbol in tmp.exp.free_symbols))]
     if not lamcheck:
         raise NotImplementedError()
 
+    def _solve_even_degree_branches(diff, symbol, t, gens):
+        """
+        Helper function to consider even degree branches of lambert
+        equation if exists.
+
+        For example:
+        Equation (a/x + exp(x/2)).diff(x) = 0 is represented
+        in logform as both of the below equations.
```

Comment:
This just does not look right. Why do we only have x changing sign in one place? Those equations together imply that log(x) == log(-x)...which we know is not true. :confused:

Reply:
@smichr Here the equation `(a/x + exp(x/2)).diff(x) = 0` is represented as two equation as
1. 2*log(x) + x/2 - log(a) = 0
2. 2*log(-x) + x/2 - log(a) = 0
which here means is **solutions of (a/x + exp(x/2)).diff(x) = solution of eq 1.- Union-solution of eq 2.**
 And not equals to `Intersection(solution of eq1, solution of eq 2)` which obviously is `S.EmptySet`.

Reply:
Please bear with me...it seems to me that you are getting the right answer for the wrong reason.

`(A/x + exp(x/2)).diff(x) = 0` implies that `A = x**2*exp(x/2)/2`. Taking log of both sides gives `log(A) = 2*log(x) + x/2`; rearrangement  and identification of constants in `a*log(b*X + c) + d*X + f` gives `a=2, b=1, c=0, d=1/2 and f=-log(A)`. The solution of that log equation is given by:
```python
>>> solve(a*log(b*x + c) + d*x + f,x)
[(a*b*LambertW(d*exp((-f + c*d/b)/a)/(a*b)) - c*d)/(b*d)]
```
It seems like you want to say that there are two solutions:
```python
a_odd_or_even = (a*b*LambertW(d*exp((-f + c*d/b)/a)/(a*b)) - c*d)/(b*d)
a_even = (a*b*LambertW(-d*exp((-f + c*d/b)/a)/(a*b)) - c*d)/(b*d)
```
I am not sure what the justification for this conclusion is. I am not saying that it is not true, just that I don't know what justifies the conclusion. In this
```
assert solve((a/x + exp(x/2)).diff(x), x) == \
            [4*LambertW(-sqrt(2)*sqrt(a)/4), 4*LambertW(sqrt(2)*sqrt(a)/4)]
```
the first solution is not real and the second is real when `a` is 2.

We can anticipate when there will be two **real** solutions:
```python
>>> eq = a*log(b*x + c) + d*x + f
>>> solve(eq.diff(x), x)
[-a/d - c/b]
```
So we will have two real solutions when neither b nor d is 0 and the 2nd derivative at that max/min is opposite in sign to the value there. Additionally, `b*x + c > 0`.
```python
>>> ext = _[0]
>>> eq.subs(x,ext)
a*log(b*(-a/d - c/b) + c) + d*(-a/d - c/b) + f
>>> eq.diff(x,2).subs(x,ext)
-a*b**2/(b*(-a/d - c/b) + c)**2
>>> cond = eq.subs(x,ext)*eq.diff(x,2).subs(x,ext) < 0
>>> cond.simplify().factor()
-d**2*(a*b*log(-a*b/d) - a*b + b*f - c*d)/(a*b) < 0
```

If there aren't two real solutions then 1 or 2 may have imaginary components.

Reply:
Yeah getting imaginary parts is a bit of a concern here (if we want only reals to be returned).
Should we just include all the solutions and let `solve` or `solveset` decide whether to take the solutions or not or add such a  check accordingly?

Reply:
That is what solve generally tries to do. I am only loosely tracking the work here but in the back of my mind I am bothered by this 'even branch' treatment and haven't seen a good demonstration, mathematically, of what/why this is being done and how the solution to those two equations leads to the right solutions and how these relate to the -1 branch of LambertW. I know it is rough going in this area -- I hobbled as much as I could from reading about the W function online and what we have currently is the fruit of that. I am glad for the progress here.